### ISSUE TITLE
IDEA: Auto generate Monoid[FunctionN]

### ISSUE THREAD

**Issue by BalmungSan** [typelevel/cats#4079]
Currently **cats** provides a `Monoid` instance for `Function0` & `Function1`: https://github.com/typelevel/cats/blob/main/kernel/src/main/scala/cats/kernel/instances/FunctionInstances.scala#L109
While these are very useful, it is not uncommon in **Scala** to have functions of higher arities like `Function2`.
And, for all I can see, the implementation of such instances is just mechanical repetition.

Thus, this issue proposes the inclusion of the generation of all the repetitive code from `trait Semigroup[Function2]` until `trait Monoid[Function22] extends Semigroup[Function22]`
AFAIK the only downside of such a proposal is the increase in the JAR size; but, personally, I think this is a good addition since not having those instances can be surprising and somehow inconsistent.

_PS: I see that we also have `Group` instances for `Function0` & `Function1`. However, I am unfamiliar with that **typeclass** and I am not totally sure if such instance is valid for higher arities and if the implementation is equally mechanical; for how the code looks like it seems so, thus I leave it open to also include those in the generator._

**Comment by armanbilge** [comment]
Re jar size, see https://github.com/typelevel/cats/pull/3871.

**Comment by satorg** [comment]
Yeah, jar size is a concern... I'm just thinking about a broader solution... I'm wondering, would it make sense to split cats-core into two parts: one is a minimalistic jar with all arities no more than 2 and the second is somewhat full-sized with all the remaining arities up to 22 (at least) and other size-costly goods? Would it work in general, wdyt?

**Comment by satorg** [comment]
@johnynek one more import for that?

**Comment by johnynek** [comment]
yes, you would need another import.

**Comment by satorg** [comment]
I mean this is one of the things we could consider for future work. Most of libraries which depend on Cats don't need high-arity syntax (I guess). But sometimes it might be useful for business-level applications to support specific DTOs or pipelines. This is where such an extension module may come handy.

**Comment by joroKr21** [comment]
This could also be part of Kittens: https://github.com/typelevel/kittens/blob/master/core/src/main/scala/cats/derived/function.scala
